workflow/controller/pod_cleanup.go

```diff
@@ -1,22 +1,142 @@
+}
+
+// @podGC is used to generate the label selector @labelPodCompleted is used to determine if
+// we should add a common.LabelKeyCompleted=true label
+func getLabelSelector(wfName string, podGC *wfv1.PodGC, labelCompleted bool) metav1.LabelSelector {
+	// upper limit on memory usage set
+	labels := make(map[string]string)
+	labels[common.LabelKeyWorkflow] = wfName
+	if labelCompleted {
+		labels[common.LabelKeyCompleted] = "true"
+	}
+
+	expressions := []metav1.LabelSelectorRequirement{}
+
+	if podGC != nil && podGC.LabelSelector != nil {
+		for key, value := range podGC.LabelSelector.MatchLabels {
+			labels[key] = value
+		}
+		expressions = append(expressions, podGC.LabelSelector.MatchExpressions...)
+	}
+	return metav1.LabelSelector{MatchLabels: labels, MatchExpressions: expressions}
+}
+
+func (woc *wfOperationCtx) runImmediateCleanup(pod *apiv1.Pod, podGC *wfv1.PodGC, workflowPhase wfv1.WorkflowPhase) error {
+	/// we have finished the cleanup previously
+	if woc.finishedCleanup {
+		return nil
+	}
+
+	log := woc.log.WithField("podName", pod.Name)
+	log.Infof("Running PodGC strategy PodGCOnWorkflowCompletion for %s", pod.Name)
+	pods := woc.controller.kubeclientset.CoreV1().Pods(pod.Namespace)
+	wfName := woc.execWf.ObjectMeta.Name
+	podGCSelector, err := podGC.GetLabelSelector()
+	if err != nil {
+		return err
+	}
+	woc.rateLimiter.Wait()
+	// check if we need to label the pod as completed first and do so if we must
+	if determinePodCleanupAction(podGCSelector, pod.Labels, wfv1.PodGCOnPodCompletion, workflowPhase, pod.Status.Phase) == labelPodCompleted {
+		_, err := pods.Patch(
+			context.Background(),
+			pod.Name,
+			types.MergePatchType,
+			[]byte(`{"metadata": {"labels": {"workflows.argoproj.io/completed": "true"}}}`),
+			metav1.PatchOptions{},
+		)
+		if err != nil {
+			log.Errorf("was not able to patch pods with completed as true for pods %s in workflow %s due to %s hence abandoning deletion", pod.Name, woc.execWf.ObjectMeta.Name, err)
+			return err
+		}
+	}
+
+	var lselector metav1.LabelSelector
+	if woc.wf.Status.Fulfilled() {
+		log.Info("workflow is fulfilled, deleting everything which may be deleted as per PodGC.Strategy.LabelSelector")
+		lselector = getLabelSelector(wfName, podGC, false)
+	} else {
+		log.Info("deleting existing pods marked as completed")
+		lselector = getLabelSelector(wfName, podGC, true)
+	}
+	selector, err := metav1.LabelSelectorAsSelector(&lselector)
+	if err != nil {
+		log.Errorf("was not able to generate a selector due to %s hence abandoning deletion", err)
+		return nil
+	}
+
+	selectorString := selector.String()
+
+	// optimisation when function is called with a fulfilled workflow
+	// lets patch everything up
+	if woc.wf.Status.Fulfilled() {
+		markDone := true
+		podList, err := pods.List(context.Background(), metav1.ListOptions{LabelSelector: selectorString})
+		if err != nil {
```

Comment:
Is it safe to continue execution after this error or should we retry the whole delete cycle?

Reply:
We do actually retry, but I did forget one thing, markDone = false should be stated in that branch. 
This enables the retry to happen when the next clean up call is made. 

Addressed.

Reply:
A nit on the use of `else` from [Effective Go
](https://go.dev/doc/effective_go):
```
In the Go libraries, you'll find that when an if statement doesn't flow into the next statement—that is, the body ends in break, continue, goto, or return—the unnecessary else is omitted.
```

I am quite new to Go but i seldom see the use of else(to the level that it is avoided) in other code bases, especially in the roam of err checking. For me the user here dont feel very go @isubasinghe  but it might be because I am new to it.

Reply:
I think else is unavoidable here unless you introduce a goto statement or `if err == nil` or repeat some code. To me, else still is the option that makes the most sense here. 

But I think that the text there just states when the if statement doesn't flow into the next statement that is the recommended approach, in this case here it does flow into the next statement in both branches, so that rule doesn't apply here.

Reply:
You could factor out into a helper function if you really want to avoid the else. Something like `patchSelectedPods(ctx context.Context, labelSelector string) error`

Reply:
Ah yes that works, addressed.